M2/Macaulay2/packages/Varieties/SheafMaps.m2

```diff
@@ -0,0 +1,890 @@
+    deg := if opts.Degree =!= null then opts.Degree else min flatten degrees source phi;
+    new SheafMap from {
+        symbol source => F,
+        symbol target => G,
+        symbol degree => deg,
+        symbol map => phi,
+        symbol cache => new CacheTable
+        }
+    )
+
+-- when phi is constructed by truncation >= d
+map(CoherentSheaf, CoherentSheaf, Matrix, ZZ) := SheafMap => opts -> (G, F, phi, d) -> (
+    newPhi := inducedMap(module G, target phi) * phi;
+    map(G, F, newPhi, Degree => d))
+map(CoherentSheaf, CoherentSheaf, Matrix, InfiniteNumber) := SheafMap => opts -> (G, F, phi, d) -> (
+    if d === -infinity then map(G, F, phi) else error "unexpected degree for map of sheaves")
+-- TODO: support map(F, F, 1) and map(F, G, 0) for identity and zero maps
+
+sheafMap = method()
+sheafMap Matrix      := SheafMap =>  phi     -> map(sheaf target phi, sheaf source phi, phi)
+sheafMap(Matrix, ZZ) := SheafMap => (phi, d) -> map(sheaf target phi, sheaf source phi, truncate(d, phi), d)
+
+isWellDefined SheafMap := f -> (
+    (G, F) := (target f, source f);
+    X := variety f;
+    d := degree f;
+    all({ G, F, matrix f }, isWellDefined)
+    -- data type checks
+    and condition(set keys f === set { symbol source, symbol target, symbol degree, symbol map, symbol cache },
+	"the hash table does not have the expected keys")
+    and condition(
+	instance(f.source, CoherentSheaf) and
+	instance(f.target, CoherentSheaf) and
+	instance(f.cache, CacheTable) and
+	instance(f.map, Matrix) and
+	instance(f.degree, ZZ),
+	"the hash table does not have the expected values")
+    -- mathematical checks
+    and condition(ring f === ring X,
+	"underlying matrix and variety do not have the same ring")
+    and condition(same {X, variety F, variety G},
+	"underlying variety does not match that of the source and target")
+    and condition(not isProjective X or isHomogeneous matrix f,
+	"underlying matrix of a map of coherent sheaf on a projective variety should be homogeneous")
+    and condition(G === sheaf(X, target matrix f),
+	"target of the sheaf map does not match the target of the underlying matrix")
+    and condition(F  == sheaf(X, source matrix f),
+	"source of the sheaf map does not match the source of the underlying matrix")
+    and condition(d >= min flatten degrees F, -- maybe not strictly necessary
+	"expected the degree of the sheaf map to be at least as high as the degrees of the source")
+    and condition(try ( isWellDefined map(module G, truncate(d, module F), matrix f) ) else false,
+	"expected the matrix to induce a map between a truncation of the underlying modules")
+    )
+
+-- basic methods
+source  SheafMap := CoherentSheaf => f -> f.source
+target  SheafMap := CoherentSheaf => f -> f.target
+variety SheafMap := Variety       => f -> f.source.variety
+ring    SheafMap := Ring          => f -> f.map.ring
+matrix  SheafMap := Matrix => opts -> f -> f.map
+degree  SheafMap := ZZ => f -> f.degree
```

Comment:
Calling this numerical invariant of a ```SheafMap``` seems problematic.  As I understand it, it is NOT the degree of the map rather the degree at which the source module was truncated to defined the morphism.

Reply:
That's a fair point. I'll change all uses of `degree psi` to `psi.degree` and comment out this method for now, but perhaps we should discuss this at the next meeting.

Reply:
Ah, actually, this seems to be used quite a while and might cause git conflicts. I agree that we should change this, but let's wait until we meet.